File: src/commands/devCommands.ts
// src/commands/devCommands.ts
import { core } from "@tauri-apps/api";
import type { ActiveSessionInfo, DevStats, DevBoxInfo } from "../types";

export const devCommands = {
    async createSession(sessionName: string, boxId: string): Promise<void> {
        return await core.invoke("create_session", {
            sessionName,
            boxId
        });
    },

    async getAllSessions(): Promise<ActiveSessionInfo[]> {
        return await core.invoke("get_all_sessions");
    },

    async getSessionInfo(sessionName: string): Promise<ActiveSessionInfo | null> {
        return await core.invoke("get_dev_session_info", {
            sessionName
        });
    },

    async stopSession(sessionName: string): Promise<void> {
        return await core.invoke("stop_session", {
            sessionName
        });
    },

    async clearAllSessions(): Promise<void> {
        return await core.invoke("clear_all_sessions");
    },

    async hasAnySessions(): Promise<boolean> {
        return await core.invoke("has_any_sessions");
    },

    async getAvailableDevBoxes(): Promise<DevBoxInfo[]> {
        return await core.invoke("get_available_dev_boxes");
    },

    async getDevStats(): Promise<DevStats> {
        return await core.invoke("get_dev_stats");
    },

    async validateSessionName(sessionName: string): Promise<boolean> {
        try {
            const result: boolean = await core.invoke("validate_session_name", {
                sessionName
            });
            console.log(`🔍 Backend availability check for "${sessionName}":`, result);
            return result;
        } catch (error) {
            console.error(`❌ Session name validation error for "${sessionName}":`, error);
            return false;
        }
    },

    async validateSessionKey(sessionKeyHex: string): Promise<boolean> {
        return await core.invoke("validate_session_key", {
            sessionKeyHex
        });
    },
};

// Session helpers remain the same
export const sessionHelpers = {
    validateSessionNameLocally(sessionName: string): { isValid: boolean; message?: string } {
        if (!sessionName || sessionName.trim().length === 0) {
            return { isValid: false, message: "Session name cannot be empty" };
        }

        if (sessionName.length > 30) {
            return { isValid: false, message: "Session name cannot exceed 30 characters" };
        }

        const pattern = /^[a-z0-9\-]{1,30}$/;
        if (!pattern.test(sessionName)) {
            return {
                isValid: false,
                message: "Session name must be 1-30 characters, lowercase letters, numbers, and hyphens only"
            };
        }

        return { isValid: true };
    },

    generateSessionName(boxName: string, attempt: number = 0): string {
        let baseName = boxName
            .toLowerCase()
            .replace(/[^a-z0-9\-]/g, '-')
            .replace(/-+/g, '-')
            .replace(/^-|-$/g, '')
            .substring(0, attempt > 0 ? 25 : 30);

        if (attempt > 0) {
            const suffix = `-${attempt}`;
            baseName = baseName.substring(0, 30 - suffix.length) + suffix;
        }

        return baseName || 'dev-session';
    }
};

File: src-tauri/src/commands/dev_commands.rs
// src/commands/dev_commands.rs

use crate::models::{ActiveSessionInfo, DevStats};
use crate::states::{DevBoxInfo, DevState};
use tauri::State;

#[tauri::command]
pub async fn create_session(
    session_name: String,
    box_id: String,
    dev_state: State<'_, tokio::sync::Mutex<DevState>>,
) -> Result<(), String> {
    let dev_state_guard = dev_state.lock().await;
    dev_state_guard
        .create_session(session_name, box_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_all_sessions(
    dev_state: State<'_, tokio::sync::Mutex<DevState>>,
) -> Result<Vec<ActiveSessionInfo>, String> {
    let dev_state_guard = dev_state.lock().await;
    dev_state_guard
        .get_all_sessions()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_dev_session_info(
    session_name: String,
    dev_state: State<'_, tokio::sync::Mutex<DevState>>,
) -> Result<Option<ActiveSessionInfo>, String> {
    let dev_state_guard = dev_state.lock().await;
    dev_state_guard
        .get_session_info(&session_name)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn stop_session(
    session_name: String,
    dev_state: State<'_, tokio::sync::Mutex<DevState>>,
) -> Result<(), String> {
    let dev_state_guard = dev_state.lock().await;
    dev_state_guard
        .stop_session(session_name)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn clear_all_sessions(
    dev_state: State<'_, tokio::sync::Mutex<DevState>>,
) -> Result<(), String> {
    let dev_state_guard = dev_state.lock().await;
    dev_state_guard
        .clear_all_sessions()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn has_any_sessions(
    dev_state: State<'_, tokio::sync::Mutex<DevState>>,
) -> Result<bool, String> {
    let dev_state_guard = dev_state.lock().await;
    dev_state_guard
        .has_any_sessions()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_available_dev_boxes(
    dev_state: State<'_, tokio::sync::Mutex<DevState>>,
) -> Result<Vec<DevBoxInfo>, String> {
    let dev_state_guard = dev_state.lock().await;
    dev_state_guard
        .get_available_dev_boxes()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_dev_stats(
    dev_state: State<'_, tokio::sync::Mutex<DevState>>,
) -> Result<DevStats, String> {
    let dev_state_guard = dev_state.lock().await;
    dev_state_guard
        .get_dev_stats()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn validate_session_name(
    session_name: String,
    dev_state: State<'_, tokio::sync::Mutex<DevState>>,
) -> Result<bool, String> {
    let dev_state_guard = dev_state.lock().await;

    // Use the dedicated public method which handles both validation and the database check.
    dev_state_guard
        .is_session_name_available(&session_name)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn validate_session_key(
    session_key_hex: String,
    dev_state: State<'_, tokio::sync::Mutex<DevState>>,
) -> Result<bool, String> {
    let dev_state_guard = dev_state.lock().await;
    match dev_state_guard.validate_session_key(&session_key_hex) {
        Ok(_) => Ok(true),
        Err(_) => Ok(false),
    }
}
